tests/unit/Core/Checkout/Cart/Rule/LineItemPerItemQuantityRuleTest.php

```diff
@@ -0,0 +1,258 @@
+        static::assertArrayHasKey('operator', $constraints);
+
+        static::assertEquals(new Type(type: 'int'), $constraints['quantity'][1]);
+        static::assertEquals(new Choice($operators), $constraints['operator'][1]);
+    }
+
+    #[DataProvider('getMatchingValues')]
+    public function testLineItemPerQuantityRuleMatchingCartScope(bool $expected, int $cartQuantity, int $quantityValue, string $operator): void
+    {
+        $cart = new Cart('test');
+        $cart->add(new LineItem('test1', 'product', null, $cartQuantity));
+
+        $scope = new CartRuleScope(
+            $cart,
+            $this->createMock(SalesChannelContext::class),
+        );
+        $this->rule->assign([
+            'quantity' => $quantityValue,
+            'operator' => $operator,
+        ]);
+
+        $isMatching = $this->rule->match($scope);
+
+        static::assertSame($expected, $isMatching);
+    }
+
+    #[DataProvider('getMatchingValues')]
+    public function testLineItemPerQuantityRuleMatchingLineItemScope(bool $expected, int $cartQuantity, int $quantityValue, string $operator): void
+    {
+        $lineItem = new LineItem('test1', 'product', null, $cartQuantity);
+
+        $scope = new LineItemScope(
+            $lineItem,
+            $this->createMock(SalesChannelContext::class),
+        );
+        $this->rule->assign([
+            'quantity' => $quantityValue,
+            'operator' => $operator,
+        ]);
+
+        $isMatching = $this->rule->match($scope);
+
+        static::assertSame($expected, $isMatching);
+    }
+
+    public function testEmptyCartShouldReturnFalse(): void
+    {
+        $cart = new Cart('test');
+
+        $scope = new CartRuleScope(
+            $cart,
+            $this->createMock(SalesChannelContext::class),
+        );
+        $this->rule->assign([
+            'quantity' => 5,
+            'operator' => Rule::OPERATOR_EQ,
+        ]);
+
+        $isMatching = $this->rule->match($scope);
+
+        static::assertFalse($isMatching);
+    }
+
+    public function testUnhandledScopeShouldReturnFalse(): void
+    {
+        $scope = new TestRuleScope(
+            $this->createMock(SalesChannelContext::class),
+        );
+
+        $this->rule->assign([
+            'quantity' => 5,
+            'operator' => Rule::OPERATOR_EQ,
+        ]);
+
+        $isMatching = $this->rule->match($scope);
+
+        static::assertFalse($isMatching);
+    }
+
+    public function testUnsetQuantityShouldReturnFalse(): void
+    {
+        $scope = new LineItemScope(
+            new LineItem('test1', 'product', null, 1),
+            $this->createMock(SalesChannelContext::class),
+        );
+
+        $this->rule->assign([
+            'operator' => Rule::OPERATOR_EQ,
+        ]);
+
+        $isMatching = $this->rule->match($scope);
+
+        static::assertFalse($isMatching);
+    }
+
+    public static function getMatchingValues(): \Generator
+    {
+        yield 'EQ - true' => [
+            true,
```

Comment:
keys for better readability:
```suggestion
            'expected' => true,
```

Reply:
not only for readability, afaik it is deprecated to use list arrays